### PR TITLE
[HOP] Fixes for zero-length scan dim

### PR DESCRIPTION
**Pull request by bohnstingl**
Currently, `scan` as well as `associative_scan` require the scan dimension of the tensor to be scanned to be > 0. These gaps have been reported in https://github.com/pytorch/pytorch/issues/186107 as well as in https://github.com/pytorch/pytorch/issues/186465. Having a scan dimension being equal to 0 is supported in JAX and is a valid usecase if masking is used on the tensor to be scanned.

This PR enables this behavior by generalizing the `first_slice_copy` function. If the scan length is 0, then a new empty tensor with the appropriate shape is returned. This approach has the benefit of enabling the zero-length case for `scan` as well as for `associative_scan` at the same time.

The testcases reported in https://gist.github.com/zou3519/80f188f5d31e2adab0e089cffc644b1c#file-scan_gaps-py-L465 and in https://gist.github.com/zou3519/4e6ed27b1c1392890905f6ee14902cd4#file-torch_jax_associative_scan_cases-py-L228 are passing with this fix. In addition, I've also added dedicated testcases.

This change was authored with the assistance of an AI coding assistant (Claude).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @ydwu4 @zou3519

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section

File: torch/_higher_order_ops/associative_scan.py [hunk 1]
```diff
@@ -785,7 +794,7 @@ def expand_masks(mask):
             zeros_mask = compute_helper_tril_mask(-1)
 
             # 5.1) Repeat the elements of bwys to form the square matrix
-            y_mat = bwys.unsqueeze(dim).repeat_interleave(scan_length, dim)
+            y_mat = bwys.unsqueeze(dim).repeat_interleave(scan_length, dim).clone()
```
Comment (ydwu4) [inline torch/_higher_order_ops/associative_scan.py]: why we need the clone?

File: torch/_higher_order_ops/associative_scan.py [hunk 2]
```diff
@@ -102,9 +104,6 @@ def __call__(self, combine_fn, xs, additional_inputs):
 
     # pyrefly: ignore [bad-override]
     def gen_schema(self, combine_fn, xs, additional_inputs):
-        from torch._higher_order_ops.schema import HopSchemaGenerator
```
Comment (ydwu4) [inline torch/_higher_order_ops/associative_scan.py:105]: these are purely refactor, right?

File: torch/_higher_order_ops/scan.py [hunk 3]
```diff
@@ -347,6 +345,28 @@ def _scan(init, xs):
...
+                )
+            out_tensor_mask = get_tensor_mask(out_proto)
+            out_proto_masked = mask_list(out_tensor_mask, out_proto)
+            outs = [
+                torch.empty([0] + list(e.size()), dtype=e.dtype, device=e.device)
+                for e in out_proto_masked
+            ]
+            outs_expanded = [
+                outs.pop(0) if out_m else None for out_m in out_tensor_mask
+            ]
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py]: this looks awkard. how jax handles this?

File: torch/_higher_order_ops/associative_scan.py (hunk 2 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/associative_scan.py:105]: Yeah, this was a refactor because of the Claude review above, see In-function imports. However, I over corrected. I moved those changes back to in-function imports as they were before.

File: torch/_higher_order_ops/associative_scan.py (hunk 1 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/associative_scan.py]: No, I don't think this is needed, removed.

File: torch/_higher_order_ops/scan.py (hunk 3 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: I think JAX is not directly invoking a forward call, but rather infers the shapes from the tracing. However, on a second look, we could greatly simplify this branch and consolidate it with the already existing behavior. I refactored.

File: torch/_higher_order_ops/scan.py [hunk 4]
```diff
@@ -383,6 +382,16 @@ def _scan(init, xs):
             torch.ones_like(e, dtype=torch.int64).unsqueeze(0) for e in out_0_masked
         ]
 
+        if num_elems == 0:
+            # outs: list[Tensor] = [
```
Comment (aorenste) [inline torch/_higher_order_ops/scan.py]: [cleanup] This commented-out block is stale. `outs` is already built as `torch.empty([0, *e.size()])` just above (scan.py:373-380, since `num_elems == 0` here), and the branch correctly reuses it via `outs.pop(0)` below. The comment describes a separate construction that never runs and is misleading -- remove lines 386-389.

File: torch/_higher_order_ops/utils.py [hunk 5]
```diff
@@ -947,8 +947,15 @@ def get_dummy_aot_autograd_config():
...
-# Slices off the first element of a given dimension
+# Slices off the first element of a given dimension. Used to build a
+# representative single slice for subgraph tracing of control-flow ops.
+# When `dim` has size 0 (e.g. a zero-length scan) there is no real slice to
+# take, so return an uninitialized tensor with `dim` removed.
 def first_slice_copy(t: torch.Tensor, dim: int = 0) -> torch.Tensor:
+    if t.shape[dim] == 0:
+        shape = list(t.shape)
+        del shape[dim]
+        return t.new_empty(shape)
```
Comment (aorenste) [inline torch/_higher_order_ops/utils.py]: [cleanup] `new_empty` returns uninitialized memory, and this fabricated slice is fed through the real `combine_fn` -- in eager `generic_scan` (scan.py:355) and in the reference `_fake_scan` (scan.py:1140) used by the new tests -- to infer output metadata. `new_zeros` keeps NaN/inf out of a `combine_fn` that validates its inputs (e.g. a finiteness check or `torch._check`), at negligible cost since the slice is empty along the removed dim. Fixing it here covers every fabricated-slice call site at once. (The earlier @claude review flagged this for `generic_scan`; it also affects `_fake_scan`.)

```suggestion
        return t.new_zeros(shape)
```

**Comment by bohnstingl** [comment]
@ydwu4 @aorenste @zou3519 if there any chance that I can get some eyes on this PR?

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

[cleanup] `scan._validate_input` now rejects `xs` leaves with mismatched scan-dim sizes (scan.py:183). This is a behavioral tightening: previously such inputs either raised `IndexError` mid-scan (a leaf shorter than `xs[0]`) or silently truncated to `xs[0]`'s length (a leaf longer). The upfront error is clearly better, but `associative_scan._validate_input` gets no equivalent check (the earlier @claude review noted the same asymmetry), so a mismatched-size `associative_scan` still fails with a confusing downstream shape error rather than this clear message. Low priority.

File: torch/_higher_order_ops/utils.py (hunk 5 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/utils.py]: Integrated.

File: torch/_higher_order_ops/scan.py (hunk 4 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Cleanup integrated.

**Comment by bohnstingl** [comment]
@aorenste could you please take another look before I merge?

File: torch/_higher_order_ops/scan.py [hunk 6]
```diff
@@ -340,13 +337,15 @@ def gen_schema(
 def generic_scan(operator, init, xs, dim=0, additional_inputs=()):
     def _scan(init, xs):
         """Perform scan on `elems` using `elems_init."""
-        carry = init
+        carry_orig = init
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py]: avoid accidental refactoring like this.

**Review by ydwu4 — COMMENTED** [review]
Can give a one-sentence summary on what's the semantic of scan forward and backward when scanned dim is zero in the doc string? What's jax semantic?

File: torch/_higher_order_ops/scan.py (hunk 6 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Good catch. This change originated during the rework that I did, It is not necessary and I reverted it.

**Comment by bohnstingl** [comment]
> Can give a one-sentence summary on what's the semantic of scan forward and backward when scanned dim is zero in the doc string? What's jax semantic?

@ydwu4 JAX doesn't really state anything specific about the zero-length inputs for neither the forward nor the backward. However, I added a sentence in the both the `scan` and the `associative_scan` docstring.

**Comment by aorenste** [comment]
Looks good to me - @ydwu4 Any concerns?

File: torch/_higher_order_ops/scan.py [hunk 7]
```diff
@@ -122,6 +123,7 @@ def scan(
             the final carry of the scan operation with same pytree structure as init.
         out (torch.Tensor or pytree with tensor leaves),
             each tensor leaf is a stacked output along first dim, where each slice is the output of a scan iteration.
+            If the scan dimension has size 0, ``final_carry`` equals ``init``, each output leaf has size 0 along ``dim``, and all input gradients are zero.
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py]: just to confirm, can write some jax's code and see if this behavior is consistent?

Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: This was a worthwhile exercise. The docstring should have actually said that the gradients are the identity. The  final_carry equals the init and thus the gradient of the final_carry w.r.t. the init is the identity. I corrected the docstring and also put together a [gist](https://gist.github.com/bohnstingl/81676cc89256d31a1a1f906d46a66c87#file-jax_pt_zero-length) to illustrate this empirically. The output on my system is:
```
JAX version:     0.10.2
PyTorch version: 2.14.0a0+git123f92c

============================================================
Section 1: torch.scan vs jax.lax.scan
============================================================
  PASS        zero_length
  PASS        gradient zero_length (scan.py:126 docstring check)

============================================================
Section 2: torch.associative_scan vs jax.lax.associative_scan
============================================================
  PASS        non_associative_subtract
  PASS        empty_scan_dimension
  PASS        gradient empty_scan_dimension

============================================================
Section 3: gradient for size-0 scan dimension (docstring check)
============================================================
  PASS        gradient scan   scalar init, xs (0,) [d(carry)/d(init)]
  PASS        gradient scan   vector init, xs (0,3) [d(carry)/d(init)]
  PASS        gradient assoc  empty xs (0,) [d(out.sum())/d(xs)]
```

**Comment by bohnstingl** [comment]
@ydwu4 The gist above compares not only the shapes, but also the actual values of the tensors from JAX and from PT, see https://gist.github.com/bohnstingl/81676cc89256d31a1a1f906d46a66c87#file-jax_pt_zero-length-L120. An empty tensor with a zero dimension is always represented literally as an empty array with just the metadata. For example
```
torch.empty((0, 1, 2, 3)) -> tensor([], size=(0, 1, 2, 3))
```

**Review by ydwu4 — APPROVED** [review]
